From cfd76505d716174ec1c4a182bdc9c2deae015841 Mon Sep 17 00:00:00 2001
From: feildmaster <admin@feildmaster.com>
Date: Thu, 10 Oct 2013 17:54:29 -0500
Subject: [PATCH] Force item data to use a tag name. Fixes BUKKIT-4809

The recent Minecraft update rendered the
e20e50f85083dc53cb5456254bcf5781ef750daa fix incorrect by adding a
compound name to the base tag in some code. This fix changes all uses
of tag changes to explicitly use a name.
---
 e6087ea3fb84381e44efa981283cd33a77b981ec.patch     | 72 ++++++++++++++++++++++
 src/main/java/net/minecraft/server/ItemStack.java  | 14 +++--
 .../craftbukkit/inventory/CraftItemStack.java      |  9 +--
 3 files changed, 85 insertions(+), 10 deletions(-)
 create mode 100644 e6087ea3fb84381e44efa981283cd33a77b981ec.patch

diff --git a/e6087ea3fb84381e44efa981283cd33a77b981ec.patch b/e6087ea3fb84381e44efa981283cd33a77b981ec.patch
new file mode 100644
index 0000000..ed92a70
--- /dev/null
+++ b/e6087ea3fb84381e44efa981283cd33a77b981ec.patch
@@ -0,0 +1,72 @@
+From e6087ea3fb84381e44efa981283cd33a77b981ec Mon Sep 17 00:00:00 2001
+From: feildmaster <admin@feildmaster.com>
+Date: Thu, 10 Oct 2013 17:54:29 -0500
+Subject: [PATCH] Force item data to use a tag name. Fixes BUKKIT-4809
+
+The recent Minecraft update rendered the
+e20e50f85083dc53cb5456254bcf5781ef750daa fix incorrect by adding a
+compound name to the base tag in some code. This fix changes all uses
+of tag changes to explicitly use a name.
+---
+ src/main/java/net/minecraft/server/ItemStack.java                | 9 +++++++--
+ .../java/org/bukkit/craftbukkit/inventory/CraftItemStack.java    | 9 +++++----
+ 2 files changed, 12 insertions(+), 6 deletions(-)
+
+diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
+index 69f81ca..808860e 100644
+--- a/src/main/java/net/minecraft/server/ItemStack.java
++++ b/src/main/java/net/minecraft/server/ItemStack.java
+@@ -117,8 +117,8 @@ public void c(NBTTagCompound nbttagcompound) {
+         }
+ 
+         if (nbttagcompound.hasKey("tag")) {
+-            // CraftBukkit - clear name from compound and make defensive copy as this data may be coming from the save thread
+-            this.tag = (NBTTagCompound) nbttagcompound.getCompound("tag").clone().setName("");
++            // CraftBukkit - make defensive copy as this data may be coming from the save thread
++            this.tag = (NBTTagCompound) nbttagcompound.getCompound("tag").clone();
+         }
+     }
+ 
+@@ -334,6 +334,11 @@ public NBTTagList getEnchantments() {
+     }
+ 
+     public void setTag(NBTTagCompound nbttagcompound) {
++        // CraftBukkit start - Set compound name to "tag," remove discrepancy
++        if (nbttagcompound != null) {
++            nbttagcompound.setName("tag");
++        }
++        // CraftBukkit end
+         this.tag = nbttagcompound;
+     }
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+index c70d41b..51d5beb 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+@@ -191,10 +191,11 @@ static boolean makeTag(net.minecraft.server.ItemStack item) {
+         if (item == null) {
+             return false;
+         }
+-        if (item.tag != null) {
+-            return true;
++
++        if (item.tag == null) {
++            item.setTag(new NBTTagCompound("tag"));
+         }
+-        item.tag = new NBTTagCompound();
++
+         return true;
+     }
+ 
+@@ -350,7 +351,7 @@ public static boolean setItemMeta(net.minecraft.server.ItemStack item, ItemMeta
+             return false;
+         }
+ 
+-        NBTTagCompound tag = new NBTTagCompound();
++        NBTTagCompound tag = new NBTTagCompound("tag");
+         item.setTag(tag);
+ 
+         ((CraftMetaItem) itemMeta).applyToItem(tag);
+-- 
+1.8.4
+
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index c21d8df..3f2b359 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -118,12 +118,9 @@ public final class ItemStack {
         }
 
         if (nbttagcompound.hasKey("tag")) {
-            // CraftBukkit - clear name from compound and make defensive copy as this data may be coming from the save thread
-            this.tag = (NBTTagCompound) nbttagcompound.getCompound("tag").clone().setName("");
-
-            // EMC - No repair cost on EMC.
-            if (this.tag.hasKey("RepairCost")) { this.tag.setInt("RepairCost", 0); }
-
+            // CraftBukkit - make defensive copy as this data may be coming from the save thread
+            this.tag = (NBTTagCompound) nbttagcompound.getCompound("tag").clone();
+            if (this.tag.hasKey("RepairCost")) { this.tag.setInt("RepairCost", 0); } // EMC
         }
     }
 
@@ -363,6 +360,11 @@ public final class ItemStack {
     }
 
     public void setTag(NBTTagCompound nbttagcompound) {
+        // CraftBukkit start - Set compound name to "tag," remove discrepancy
+        if (nbttagcompound != null) {
+            nbttagcompound.setName("tag");
+        }
+        // CraftBukkit end
         this.tag = nbttagcompound;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index c70d41b..51d5beb 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -191,10 +191,11 @@ public final class CraftItemStack extends ItemStack {
         if (item == null) {
             return false;
         }
-        if (item.tag != null) {
-            return true;
+
+        if (item.tag == null) {
+            item.setTag(new NBTTagCompound("tag"));
         }
-        item.tag = new NBTTagCompound();
+
         return true;
     }
 
@@ -350,7 +351,7 @@ public final class CraftItemStack extends ItemStack {
             return false;
         }
 
-        NBTTagCompound tag = new NBTTagCompound();
+        NBTTagCompound tag = new NBTTagCompound("tag");
         item.setTag(tag);
 
         ((CraftMetaItem) itemMeta).applyToItem(tag);
-- 
1.8.4.1

